src/Native/System.Security.Cryptography.Native/pal_rsa.cpp

```diff
@@ -12,3 +31,122 @@ extern "C" RSA* DecodeRsaPublicKey(const unsigned char* buf, int32_t len)
 
     return d2i_RSAPublicKey(nullptr, &buf, len);
 }
+
+extern "C" int32_t RsaPublicEncrypt(int32_t flen, const uint8_t* from, uint8_t* to, RSA* rsa, int32_t useOaepPadding)
+{
+    int padding = useOaepPadding ? RSA_PKCS1_OAEP_PADDING : RSA_PKCS1_PADDING;
+    return RSA_public_encrypt(flen, from, to, rsa, padding);
+}
+
+extern "C" int32_t RsaPrivateDecrypt(int32_t flen, const uint8_t* from, uint8_t* to, RSA* rsa, int32_t useOaepPadding)
+{
+    int padding = useOaepPadding ? RSA_PKCS1_OAEP_PADDING : RSA_PKCS1_PADDING;
+    return RSA_private_decrypt(flen, from, to, rsa, padding);
+}
+
+extern "C" int32_t RsaSize(RSA* rsa)
+{
+    return RSA_size(rsa);
+}
+
+extern "C" int32_t RsaGenerateKeyEx(RSA* rsa, int32_t bits, BIGNUM* e)
+{
+    return RSA_generate_key_ex(rsa, bits, e, nullptr);
+}
+
+extern "C" int32_t RsaSign(int32_t type, const uint8_t* m, int32_t mlen, uint8_t* sigret, int32_t* siglen, RSA* rsa)
+{
+    if (!siglen)
+    {
+        assert(false);
+        return 0;
+    }
+
+    unsigned int unsignedSigLen = 0;
+    int32_t ret = RSA_sign(type, m, UnsignedCast(mlen), sigret, &unsignedSigLen, rsa);
+    assert(unsignedSigLen <= INT32_MAX);
+    *siglen = static_cast<int32_t>(unsignedSigLen);
+    return ret;
+}
+
+extern "C" int32_t RsaVerify(int32_t type, const uint8_t* m, int32_t mlen, uint8_t* sigbuf, int32_t siglen, RSA* rsa)
+{
+    return RSA_verify(type, m, UnsignedCast(mlen), sigbuf, UnsignedCast(siglen), rsa);
+}
+
+extern "C" void GetRsaParameters(const RSA* rsa,
+                                 BIGNUM** n,
+                                 BIGNUM** e,
+                                 BIGNUM** d,
+                                 BIGNUM** p,
+                                 BIGNUM** dmp1,
+                                 BIGNUM** q,
+                                 BIGNUM** dmq1,
+                                 BIGNUM** iqmp)
+{
+    if (!rsa || !n || !e || !d || !p || !dmp1 || !q || !dmq1 || !iqmp)
+    {
+        assert(false);
+        return;
```

Comment:
> Is what you are looking for, something like this?

Yes, either that, or declaring the managed signature to use ref for these rather than out.

The problem here is exacerbated because this function doesn't return any error code, so imagine the managed code invoked it like:
```C#
SafeRsaHandle rsa = GetHandle();
IntPtr n, e, d, p, dmp1, q, dmq1, iqmp;
GetRsaParameters(rsa, out n, out e, out d, out p, out dmp1, out q, out dmq1, out iqmp);
```
and imagine that GetHandle() returned an IsInvalid handle. Without this null setting, those IntPtrs are going to contain garbage values and the caller will be none-the-wiser.  The one call site we have to the managed GetRsaParameters does an explicit check for IsInvalid, but we could easily find ourselves missing that in a future call to this.